### PR TITLE
lots of documentation 

### PR DESCRIPTION
**Pull request by h-2**
This is not yet done, but already contains lots changes so I would be happy comment on this.

### CODE REVIEW

File: doc/howto/porting/index.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+## Paths and Namespaces
+
+For SeqAn1 and SeqAn2 the following entities are called "seqan":
+
+  * the primary namespace
+  * the include folder
+  * the CMake library name
+
+All of the above are named "seqan3" for the SeqAn3 library so coexistance should be possible on all levels.
```
Comment (eseiler) [inline doc/howto/porting/index.md]:
```suggestion
All of the above are named "seqan3" for the SeqAn3 library so coexistence should be possible on all levels.
```

File: doc/howto/porting/index.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+We discourage having any `using namespace seqan;` or `using namespace seqan3;` statements in your application and
+and suggest replacing any occurences of the former with fully qualified lookups, e.g. writing `seqan::readRecord`.
```
Comment (eseiler) [inline doc/howto/porting/index.md]:
```suggestion
and suggest replacing any occurrences of the former with fully qualified lookups, e.g. writing `seqan::readRecord`.
```

File: doc/howto/porting/index.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+  * Input and Output.
+
+## SeqAn2 features missing from SeqAn3
+
+This is a non-exhaustive list of features that are currently missing, but **will be ported**:
+
+  * Annotation file formats (GFF, GTF, VCF...)
+  * Blast tabular file formats
+  * Memory Mapping support
+  * Journalled String Tree
```
Comment (eseiler) [inline doc/howto/porting/index.md]:
```suggestion
  * Journaled String Tree
```

File: doc/howto/porting/index.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+This is a non-exhaustive list of features that we **do not plan to port:**
+
+  * The Store module
+  * Online-Search algorithms (not used in practice)
+  * The Graph module (use the lemon library instead)
+  * Uncompressed suffix arrays (no benifit over FM indexes)
```
Comment (eseiler) [inline doc/howto/porting/index.md]:
```suggestion
  * Uncompressed suffix arrays (no benefit over FM indexes)
```

File: doc/setup/library_tests/index.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+Invoke CMake (this is often referred to as the "Configure" step):
+
+```bash
+cmake ../../seqan3/test/unit -DCMAKE_BUILD_TYPE=Debug -DCMAKE_CXX_COMPILER=g++-7
+```
+
+The build type could be "Release", but stick to "Debug" for now. Specifying the compiler is optional; depending on
+your setup, you may need to give the full path.
+
+Invoke Make (this is the actual build step):
```
Comment (eseiler) [inline doc/setup/library_tests/index.md]: Should we also talk about how to build a single test instead building the complete suite? :thinking:

Comment (h-2) [reply doc/setup/library_tests/index.md]: Good point!

File: doc/about/api/index.md [hunk 6]
```diff
@@ -2,44 +2,68 @@
```
Comment (eseiler) [inline doc/about/api/index.md]: One line above:

`proofness` instead of `proofeness`

File: doc/about/api/index.md [hunk 7]
```diff
@@ -2,44 +2,68 @@
...
-     unannounced change. This is usually the case for auxiliary data structures that would be part of
-     `seqan3::detail` were they not needed to generate correct API documentation for entities in `seqan3`.
-
-All of the above cases will be distinctly visible in the documentation.
+Special rules that override the previous statement:
+  1. **Do not depend upon internal details.** If something is in a namespace or filename/path that includes the word
+     `detail` or if it is private to a class type, you are not allowed to depend upon it any way. More generally,
+     if it's not part of the *user documentation*, it's not part of the API.
+  2. **Do not depend on entities marked as "experimental" or "NOAPI"**. Major changes to the library like new modules
+     are often marked as experimental within the first minor release  they appear in. This means we do not guarantee
```
Comment (eseiler) [inline doc/about/api/index.md]:
```suggestion
     are often marked as experimental within the first minor release they appear in. This means we do not guarantee
```

File: doc/about/api/index.md [hunk 8]
```diff
@@ -2,44 +2,68 @@
...
+     stability until the next minor release happens and the experimental flag is removed. Very few entities in namespace
+     `seqan3` are permanently marked as "NOAPI" which designates them as subject to unannounced change. This is usually the case for auxiliary data structures that would be part of `seqan3::detail` were they not needed to generate
+     correct API documentation for entities in `seqan3`.
+  4. **Do not depend on the *signatures* of SeqAn APIs.** In particular you may not take the address of APIs in SeqAn
+     and you may not use metaprogramming tricks to depend on those signatures. We reserve the right to:
+     * Add new names to namespace `seqan3` and any sub-namespaces
+     * Add new member functions to types in namespace `seqan3`
+     * Add new overloads to existing functions
+     * Add new default arguments to functions and templates
+     * Change return-types of functions in compatible ways (void to anything, numeric types in a widening fashion, etc).
```
Comment (eseiler) [inline doc/about/api/index.md]:
```suggestion
     * Change return-types of functions in compatible ways (void to anything, numeric types in a widening fashion, etc.).
```

File: doc/about/customisation/index.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+# Generic interfaces
+
+Templates are a core part of generic programming in C++, because they allow instantiating the code with different
+types or constants. SeqAn's API consists mostly of templates for exactly this reason, although we try to make this
+less visible by having the template parameters deduced in many scenarios.
+
+We strive to document the exact requirements for every template parameter in the API documentation; often these
+are also enforced via C++ concepts. Wherever other types meet the given requirements, you can use these
+type as arguments – independet of whether they are standard library types, your own types or types from another
```
Comment (eseiler) [inline doc/about/customisation/index.md]:
```suggestion
type as arguments – independent of whether they are standard library types, your own types or types from another
```

File: doc/about/api/index.md [hunk 10]
```diff
@@ -2,44 +2,68 @@
...
-  2. Major changes to the library like new modules are often marked as experimental within the first minor release
-     they appear in. This means we do not guarantee stability until the next minor release happens and the experimental
-     flag is removed. (This allows us to gather more user experience with new interfaces and make minor changes if
-     necessary.)
-  3. Very few entities in namespace `seqan3` are permanently marked as "NOAPI" which designates them as subject to
-     unannounced change. This is usually the case for auxiliary data structures that would be part of
-     `seqan3::detail` were they not needed to generate correct API documentation for entities in `seqan3`.
-
-All of the above cases will be distinctly visible in the documentation.
+Special rules that override the previous statement:
```
[the remark this answers is not in the document]
Comment (h-2) [reply doc/about/api/index.md]: I have changed it to:

```
Exceptions to the previous rule:
```

ok?

**Comment by h-2** [comment]
Can you mark your comments as resolved?

**Comment by h-2** [comment]
@rrahn Do you want to have a brief look over this before we merge it? If not, no problem. Mostly the API stuff you might be interested in seeing.

**Comment by rrahn** [comment]
@h-2 I will look into it quickly in half an hour.

File: doc/about/customisation/index.md [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+Some requirements are more elaborate and depend on a specialisation or "overload" of another entity from the
+SeqAn library. We call these "customisation points" (also known as "extension points").
+
+# Customisation points
+
+One such customisation point is seqan3::to_rank. Instead of the seqan3::Semialphabet concept looking directly for
+member functions or free functions in you namespace, it always checks for a valid definition of seqan3::to_rank which
+in turn resolves to different possible implementations.
+
+The design pattern we use for customisation points is often called a "Niebloid" (after Eric Niebler); more technical
```
Comment (rrahn) [inline doc/about/customisation/index.md]: I wouldn't put this in here. These are officially called Customisation Point Objects and I would stick to it. Many people won't know what a Niebloid really is.

**Comment by h-2** [comment]
As all of the last changes are not tests, I will merge as-is.

**Pull request merged**